compiler/src/dotty/tools/dotc/transform/LazyVals.scala

```diff
@@ -129,50 +129,58 @@ class LazyVals extends MiniPhase with IdentityDenotTransformer {
+   */
   def transformLocalDef(x: ValOrDefDef)(implicit ctx: Context): Thicket = {
-      val valueInitter = x.rhs
-      val xname = x.name.asTermName
-      val holderName = LazyLocalName.fresh(xname)
-      val initName = LazyLocalInitName.fresh(xname)
-      val tpe = x.tpe.widen.resultType.widen
-
-      val holderType =
-        if (tpe isRef defn.IntClass) "LazyInt"
-        else if (tpe isRef defn.LongClass) "LazyLong"
-        else if (tpe isRef defn.BooleanClass) "LazyBoolean"
-        else if (tpe isRef defn.FloatClass) "LazyFloat"
-        else if (tpe isRef defn.DoubleClass) "LazyDouble"
-        else if (tpe isRef defn.ByteClass) "LazyByte"
-        else if (tpe isRef defn.CharClass) "LazyChar"
-        else if (tpe isRef defn.ShortClass) "LazyShort"
-        else "LazyRef"
-
-
-      val holderImpl = ctx.requiredClass("dotty.runtime." + holderType)
-
-      val holderSymbol = ctx.newSymbol(x.symbol.owner, holderName, containerFlags, holderImpl.typeRef, coord = x.pos)
-      val initSymbol = ctx.newSymbol(x.symbol.owner, initName, initFlags, MethodType(Nil, tpe), coord = x.pos)
-      val result = ref(holderSymbol).select(lazyNme.value).withPos(x.pos)
-      val flag = ref(holderSymbol).select(lazyNme.initialized)
-      val initer = valueInitter.changeOwnerAfter(x.symbol, initSymbol, this)
-      val initBody =
-        adaptToType(
-          ref(holderSymbol).select(defn.Object_synchronized).appliedTo(
-            adaptToType(mkNonThreadSafeDef(result, flag, initer, nullables = Nil), defn.ObjectType)),
-          tpe)
-      val initTree = DefDef(initSymbol, initBody)
-      val holderTree = ValDef(holderSymbol, New(holderImpl.typeRef, List()))
-      val methodBody = tpd.If(flag.ensureApplied,
-        result.ensureApplied,
-        ref(initSymbol).ensureApplied).ensureConforms(tpe)
-
-      val methodTree = DefDef(x.symbol.asTerm, methodBody)
-      ctx.debuglog(s"found a lazy val ${x.show},\nrewrote with ${holderTree.show}")
-      Thicket(holderTree, initTree, methodTree)
+    val xname = x.name.asTermName
+    val holderName = LazyLocalName.fresh(xname)
+    val initName = LazyLocalInitName.fresh(xname)
+    val tpe = x.tpe.widen.resultType.widen
+
+    // val x$lzy = new scala.runtime.LazyInt()
+    val holderImpl = defn.LazyHolder()(ctx)(tpe.typeSymbol)
+    val holderSymbol = ctx.newSymbol(x.symbol.owner, holderName, containerFlags, holderImpl.typeRef, coord = x.pos)
+    val holderTree = ValDef(holderSymbol, New(holderImpl.typeRef, List()))
+
+    val holderRef = ref(holderSymbol)
+    val getValue = holderRef.select(lazyNme.value).ensureApplied.withPos(x.pos)
+    val initialized = holderRef.select(lazyNme.initialized).ensureApplied
+
+    // def x$lzycompute(): Int = x$lzy.synchronized {
+    //   if (x$lzy.initialized()) x$lzy.value()
+    //   else x$lzy.initialize(<RHS>)
+    // }
+    val initSymbol = ctx.newSymbol(x.symbol.owner, initName, initFlags, MethodType(Nil, tpe), coord = x.pos)
```

Comment:
I would move the definition of `initName` just before this line since that's where it's used